### PR TITLE
feat(table): added column settings feature

### PR DESCRIPTION
**Pull request by peterpham-kyndryl**
## Summary

This PR is regarding the feature where the user will be able to hide/show columns and freeze a selected column. 
For more details, here is a video Robert walked us through what the feature is [https://kyndryl.sharepoint.com/:v:/r/teams/BridgeDesignSystem/Shared%20Documents/Design%20Discussion/Freezing%20First%20Column.mov?csf=1&web=1&e=bfcLt2&nav=eyJyZWZlcnJhbEluZm8iOnsicmVmZXJyYWxBcHAiOiJTdHJlYW1XZWJBcHAiLCJyZWZlcnJhbFZpZXciOiJTaGFyZURpYWxvZy1MaW5rIiwicmVmZXJyYWxBcHBQbGF0Zm9ybSI6IldlYiIsInJlZmVycmFsTW9kZSI6InZpZXcifX0%3D](url)

And Figma link https://www.figma.com/file/P3l7EMKOWmohY4aqYfGh1e/Florence-Prototypes?type=design&node-id=5003%3A42510&mode=design&t=ajNzavtg2AEy7jQz-1

## ADO Story or GitHub Issue Link

Link here (if applicable).

## Notes

- Detailed change notes
- can go here

## To Do

- Anything else
- left to do

## Checklist

- [ ] Used Conventional Commit messages as outlined in the contributing guide.
  - [ ] Noted breaking changes (if any).
- [ ] Documented/updated all props, events, slots, parts, etc with JSDoc.
  - [ ] Ran the `analyze` command to update Storybook docs.
- [ ] Added/updated Stories with controls to cover all variants.
- [ ] Ran `test` locally to address any failures.
- [ ] Added/updated the Figma link for the Story's Design tab.
- [ ] Added any new component exports to the src/index.ts file

## Testing Instructions

Provide guidance to reviewers/testers here.

## Screenshots

(if any)

### CODE REVIEW

**Comment by maulik-thaker1** [comment]
if you don't mind can you add description what all about this PR ? So that I can understand. You can use Summary and description for this. Thanks..!!

**Comment by peterpham-kyndryl** [comment]
> if you don't mind can you add description what all about this PR ? So that I can understand. You can use Summary and description for this. Thanks..!!

Sure. Updated.

File: src/components/reusable/table/story-helpers/table.settings.ts [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+          position: absolute;
+          width: 1px;
+          height: 1px;
+          padding: 0;
+          margin: -1px;
+          overflow: hidden;
+          clip: rect(0, 0, 0, 0);
+          border: 0;
+        }
+      </style>
```
Comment (maulik-thaker1) [inline src/components/reusable/table/story-helpers/table.settings.ts:280]: its better to add styles in separate .scss file. Just suggestion

File: src/components/reusable/table/table-row.ts [hunk 2]
```diff
@@ -194,7 +225,7 @@ export class TableRow extends LitElement {
         ? html`
             <kyn-td .align=${'center'} ?dense=${this.dense}>
               <kyn-checkbox
-                ?disabled=${this.disabled}
+                ?disabled=${this.disabled || this.locked}
                 .checked=${this.selected}
```
Comment (maulik-thaker1) [inline src/components/reusable/table/table-row.ts:229]: `disabled` and `locked` same ? What's the difference ?

File: src/components/reusable/table/story-helpers/column-setting.ts [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    .min-max-width-100 {
+      --kyn-td-min-width: 100px;
+      --kyn-td-max-width: 100px;
+      --kyn-td-width: 100px;
+    }
+    .min-max-width-th-100 {
+      --kyn-th-min-width: 100px;
+      --kyn-th-max-width: 100px;
+      --kyn-th-width: 100px;
+    }
```
Comment (maulik-thaker1) [inline src/components/reusable/table/story-helpers/column-setting.ts:26]: IMO here `--kyn-th-width: 100px;` is enough unless it's shrinks or override by parent. Curious why we used min, max and width same px;

File: src/components/reusable/table/story-helpers/table.settings.ts (hunk 1 above)
Comment (peterpham-kyndryl) [reply src/components/reusable/table/story-helpers/table.settings.ts:280]: Well this is just a story, not a real component. Are we trying to formalizing story components? I'm fine if we all agree on this.
But just FYI - Today I saw inline styles in Card stories too.

File: src/components/reusable/table/table-row.ts (hunk 2 above)
Comment (peterpham-kyndryl) [reply src/components/reusable/table/table-row.ts:229]: I have the description of each prop in details in code and on Storybook, too.
So, basically a disabled and locked are similar, except for that a locked row will have a different bottom-border. And locked is specially used for column setting feature.

File: src/components/reusable/table/story-helpers/column-setting.ts (hunk 3 above)
Comment (peterpham-kyndryl) [reply src/components/reusable/table/story-helpers/column-setting.ts:26]: This is a component built specially as a part of a story, not a real component. And I remember having all three to showcase a dev, asking a way to set column width; and this is one of the ways.

File: src/components/reusable/table/story-helpers/table.settings.ts (hunk 1 above)
Comment (brian-patrick-3) [reply src/components/reusable/table/story-helpers/table.settings.ts:280]: `sr-only` can be imported from Foundation/Visibility, but I think a bigger issue is the Settings button in the global filter is not used as the anchor for the Side Drawer, that would make more sense IMO. I noticed I also have to click "Settings" twice to re-open the drawer as well.

**Review by brian-patrick-3 — COMMENTED** [review]
Only had one specific comment on the Drawer/Settings button, but I am getting concerned that some of  these table features that are not baked-in (they are more patterns really at this stage) are going to be difficult for devs to integrate into their own apps. Lots of copy/paste and translating from Lit. I wonder if in a future release we could condense some of this stuff into more reusable functions/components that we could provide for a more streamlined adoption experience.

**Comment by peterpham-kyndryl** [comment]
> Only had one specific comment on the Drawer/Settings button, but I am getting concerned that some of these table features that are not baked-in (they are more patterns really at this stage) are going to be difficult for devs to integrate into their own apps. Lots of copy/paste and translating from Lit. I wonder if in a future release we could condense some of this stuff into more reusable functions/components that we could provide for a more streamlined adoption experience.

I wanted to make the open drawer as the 

> Only had one specific comment on the Drawer/Settings button, but I am getting concerned that some of these table features that are not baked-in (they are more patterns really at this stage) are going to be difficult for devs to integrate into their own apps. Lots of copy/paste and translating from Lit. I wonder if in a future release we could condense some of this stuff into more reusable functions/components that we could provide for a more streamlined adoption experience.

I wanted to use it as the anchor, but I haven't yet found a way. The global header has its own named slot, 'actions', and to use it as an anchor, it needs to be slotted in the side drawer.

**Comment by brian-patrick-3** [comment]
> I wanted to use it as the anchor, but I haven't yet found a way. The global header has its own named slot, 'actions', and to use it as an anchor, it needs to be slotted in the side drawer.

You should be able to do this for example:
```html
<kyn-global-filter>
  <kyn-side-drawer slot="actions">
    <kd-button slot="anchor">Test</kd-button>
  </kyn-side-drawer>
</kyn-global-filter>
```

**Comment by peterpham-kyndryl** [comment]
> > I wanted to use it as the anchor, but I haven't yet found a way. The global header has its own named slot, 'actions', and to use it as an anchor, it needs to be slotted in the side drawer.
> 
> You should be able to do this for example:
> 
> ```
> <kyn-global-filter>
>   <kyn-side-drawer slot="actions">
>     <kd-button slot="anchor">Test</kd-button>
>   </kyn-side-drawer>
> </kyn-global-filter>
> ```

This way it won't be a part of the global filter

**Comment by peterpham-kyndryl** [comment]
> > I wanted to use it as the anchor, but I haven't yet found a way. The global header has its own named slot, 'actions', and to use it as an anchor, it needs to be slotted in the side drawer.
> 
> You should be able to do this for example:
> 
> ```
> <kyn-global-filter>
>   <kyn-side-drawer slot="actions">
>     <kd-button slot="anchor">Test</kd-button>
>   </kyn-side-drawer>
> </kyn-global-filter>
> ```
getting accessibility errors with this
Screenshot 2024-04-30 at 9 46 02 AM

File: src/components/reusable/table/story-helpers/column-setting.ts (hunk 3 above)
Comment (maulik-thaker1) [reply src/components/reusable/table/story-helpers/column-setting.ts:26]: got it

**Review by brian-patrick-3 — APPROVED** [review]

**Pull request merged**